arch/x86/boot/tpm_early.c: code for early hashing and extending PCR

krystian-hebel opened the pull request:
Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel committed:
arch/x86/boot/tpm_early.c: code for early hashing and extending PCR

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel committed:
arch/x86/(boot/tpm_early.c -> tpm.c): move file to not duplicate code later

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel committed:
arch/x86/dom0_build.c: measure kernel and initrd before dom0 is constructed

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

miczyg1 commented:
@krystian-hebel let me know when it will be ready for review, I see WIP in the title still

krystian-hebel committed:
arch/x86/sha1.c: add file

File comes from [1] and is licensed under MIT License. Only enough
changes to make it compile under Xen and to swap endianness of result
were made to the original file.

[1] https://www.nayuki.io/page/fast-sha1-hash-implementation-in-x86-assembly

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel committed:
arch/x86/tpm.c: calculate hash on host side instead of using TPM

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel commented:
@miczyg1 ready for review.

miczyg1 commented on xen/arch/x86/tpm.c:
```diff
@@ -0,0 +1,375 @@
+    uint8_t buf[CMD_RSP_BUF_SIZE];
+};
+
+static void tpm_hash_extend12(unsigned loc, uint8_t *buf, unsigned size,
```
Why do we need two implemenations?

krystian-hebel replied:
Just so that SHA1 (and other hashing algorithms later) isn't compiled for early, 32b flat world. MBI is small enough to have it hashed by TPM in a reasonable time.

miczyg1 replied:
I see
